autopush/tests/test_integration.py

```diff
@@ -0,0 +1,629 @@
+        return result
+
+    def unregister(self, chid):
+        msg = json.dumps(dict(messageType="unregister", channelID=chid))
+        log.debug("Send: %s", msg)
+        self.ws.send(msg)
+        result = json.loads(self.ws.recv())
+        log.debug("Recv: %s", result)
+        return result
+
+    def send_notification(self, channel=None, version=None, data=None,
+                          use_header=True, status=200, ttl=200):
+        if not channel:
+            channel = random.choice(self.channels.keys())
+
+        endpoint = self.channels[channel]
+        url = urlparse.urlparse(endpoint)
+        http = None
+        if url.scheme == "https":  # pragma: nocover
+            http = httplib.HTTPSConnection(url.netloc)
+        else:
+            http = httplib.HTTPConnection(url.netloc)
+
+        if self.use_webpush:
+            headers = {
+                "Content-Type": "application/octet-stream",
+                "Content-Encoding": "aesgcm-128",
+                "Encryption": self._crypto_key,
+                "Encryption-Key": 'keyid="a1"; key="JcqK-OLkJZlJ3sJJWstJCA"',
+                "TTL": str(ttl),
+            }
+            body = data or ""
+            method = "POST"
+            status = 201
+        else:
+            if data:
+                body = "version=%s&data=%s" % (version or "", data)
+            else:
+                body = "version=%s" % (version or "")
+            if use_header:
+                headers = {"Content-Type": "application/x-www-form-urlencoded"}
+            else:
+                headers = {}
+            method = "PUT"
+
+        log.debug("%s body: %s", method, body)
+        http.request(method, url.path, body, headers)
+        resp = http.getresponse()
+        log.debug("%s Response: %s", method, resp.read())
+        eq_(resp.status, status)
+        if self.use_webpush and ttl != 0:
+            assert(resp.getheader("Location", None) is not None)
+
+        # Pull the notification if connected
+        if self.ws and self.ws.connected:
+            result = json.loads(self.ws.recv())
+            return result
+
+    def get_notification(self, timeout=0.2):
+        self.ws.settimeout(timeout)
+        try:
+            d = self.ws.recv()
+            log.debug("Recv: %s", d)
+            return json.loads(d)
+        except:
+            return None
+
+    def ping(self):
+        log.debug("Send: %s", "{}")
+        self.ws.send("{}")
+        result = self.ws.recv()
+        log.debug("Recv: %s", result)
+        eq_(result, "{}")
+        return result
+
+    def ack(self, channel, version):
+        msg = json.dumps(dict(messageType="ack",
+                              updates=[dict(channelID=channel,
+                                            version=version)]))
+        log.debug("Send: %s", msg)
+        self.ws.send(msg)
+
+    def disconnect(self):
+        self.ws.close()
+
+
+class IntegrationBase(unittest.TestCase):
+    def setUp(self):
```

Comment:
Silly question—are these `setUp` and `tearDown` methods called for each `test_` method, or just once?